### ISSUE TITLE
scheduling of notification: handler trigger time on next day; define if one shot or recurrent

### ISSUE THREAD

**Issue by RedJocker** [hatepoint/phrases-hyperskill-tests#14]
the current code for scheduling notifications through alarmManager is this:

https://github.com/hatepoint/phrases-hyperskill-tests/blob/deb6d7f3a4effb2373364c06e318d60468ad982b/Phrases/stage3/src/main/java/org/hyperskill/phrases/MainActivity.kt#L52-L65

It is difficult to test manually, but I believe that current solution is triggering only one notification and it is not handling any time before current time. So if a person is at 23:58 and pick 00:02 it will try scheduling for a time on the past of current time, because you don't handle choosing next day in certain cases.

Another thing is that I think only one notification will be scheduled and I was thinking the intention was to have something scheduled for every day on same time. If that was indeed the intention than maybe something like this should be used:

 ```
 alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, calendar.timeInMillis, 24 * 60 * 60 * 1000, pendingIntent)
 ```
 
 I'm not that familiar with AlarmManager methods, but by reading the docstring that seems like a better fit. 
 Also didn't really test this manually, but that is difficult to do because of waiting time.

**Comment by hatepoint** [comment]
Yeah, you're right. It is supposed to be a setRepeating or something like that, but for some reason I used the usual one for testing purposes. I'll change that

**Comment by hatepoint** [comment]
The only problem here is that I'm not sure how I'm supposed to test repeating notifications while keeping in mind all the pain with problems caused by time while testing

**Comment by RedJocker** [comment]
The problem will only happen if you go through the time picking routine after advancing the clock significantly. But I don't think there will be any problem if you advance it to trigger the first time, do assertions and then advance the clock to trigger a second time to do a new round of assertions.
Don't forget to compare notifications. The second notification should not be the same instance as the first since it should be a newly triggered notification.

If some time related problem does happen warn me

**Comment by hatepoint** [comment]
Reccuring notifications work, but there's a problem: Android doesn't always post notifications at exact interval, so tests _might_ fail. I'm not sure what can be done in this case though. I'll try creating a test that checks notifications which are supposed to arrive the next day and reply later

**Comment by RedJocker** [comment]
> Android doesn't always post notifications at exact interval

yes that is true, but I don't think that will be a problem with tests because we are kind of scheduling it manually on supportAlarmManager method

https://github.com/hatepoint/phrases-hyperskill-tests/blob/95a528ec1c5eb89d8bad4b03f50c7248588b6ddf/Phrases/stage4/src/test/java/org/hyperskill/phrases/internals/PhrasesUnitTest.kt#L119-L141

**Comment by hatepoint** [comment]
Can i ask you to try to implement this test? I can't seem to understand how it's done. I'm trying to use `SystemClock.setCurrentTimeMillis(System.currentTimeMillis()+86400000)` to set the clock to the next day but even though I'm sure the notification didn't arrive it still shows up as a correct answer.

```
@Test
    fun test07_checkNotificationSentOnNextDay() {
        addToDatabase(fakePhrases)

        testActivity {
            val minutesToAdd = 10
            val calendar = Calendar.getInstance()
            calendar.add(Calendar.MINUTE, minutesToAdd)
            val pickHour = calendar.get(Calendar.HOUR_OF_DAY)
            val pickMinute = calendar.get(Calendar.MINUTE)

            reminderTv.clickAndRun()
            val timePickerDialog = getLatestTimePickerDialog()

            timePickerDialog.pickTime(pickHour, pickMinute)
            shadowLooper.idleFor(minutesToAdd + 2L, TimeUnit.MINUTES) // trigger alarm

            supportForAlarmManager()

            val notification: android.app.Notification? = notificationManager.getNotification(NOTIFICATION_ID)

            val messageNotificationId =
                "Could not find notification with id 393939. Did you set the proper id?"
            assertNotNull(messageNotificationId, notification)
            notification!!

            val messageContent = "The phrase in the notification doesn't any one in the database."
            val actualContent = notification.extras.getCharSequence(Notification.EXTRA_TEXT)?.toString()
            assertTrue(messageContent, actualContent in fakePhrases)

            SystemClock.setCurrentTimeMillis(System.currentTimeMillis()+86400000)

            val notification2: android.app.Notification? = notificationManager.getNotification(NOTIFICATION_ID)

            val messageNotificationId2 =
                "Could not find notification with id 393939. Did you set the proper id?"
            assertNotNull(messageNotificationId2, notification2)
            notification!!

            val messageContent2 = "The phrase in the notification doesn't any one in the database."
            val actualContent2 = notification.extras.getCharSequence(Notification.EXTRA_TEXT)?.toString()
            assertTrue(messageContent2, actualContent2 in fakePhrases)
        }
    }
```

It's a pretty straightforward way to check but I don't really have any other ideas for now

**Comment by RedJocker** [comment]
Yeah I see some problems, but I think I can make what you mean to do

**Comment by hatepoint** [comment]
@RedJocker I've used setExactAndAllowWhileIdle and tests still passed, meaning that it still registers no new notification as a correct solution :(

**Comment by RedJocker** [comment]
good catch, tests were schedulling one same alarm twice. Already sent a request with update

**Comment by hatepoint** [comment]
probably because it seems to send the same notification again and again, I will try to fix that and see whether that solves the problem

**Comment by RedJocker** [comment]
Sorry man, I hurried on this because I was going out. I won't be able to see this for the next 4-5 hours. Check the logic on supportForAlarmManager to see if why if it is schedulling a new alarm on the next day

**Comment by RedJocker** [comment]
I figured it out, on last update I was removing all alarms that had triggerAtTime lower then currentTime thinking that a repeating alarm would add a new ScheduledAlarm to the list at a later moment, but it does not. 
If a ScheduledAlarm has interval then it just has a property interval and no other instance is going to be added later.

So to solve this I created a way to copy the ScheduledAlarm and added a new instance with triggerAtTime = triggerAtTime + interval, if interval > 0.
